src/modules/schnorrsig/main_impl.h

```diff
@@ -236,4 +236,163 @@ int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned cha
            secp256k1_fe_equal_var(&rx, &r.x);
 }
 
+static int compute_midhash(const secp256k1_context* ctx, unsigned char *midhash, unsigned char **sig64, unsigned char *aggsig, unsigned char **msg32, secp256k1_xonly_pubkey *pubkey, size_t n_sigs) {
+    secp256k1_sha256 hash;
+    uint32_t i;
+
+    VERIFY_CHECK((aggsig != NULL) != (sig64 != NULL));
+
+    /* TODO: use actual tagged hash */
+    secp256k1_sha256_initialize(&hash);
+    /* z_i = int(hash_{HalfAggregation}(r_1 || pk_1 || m_1 || ... || r_n || pk_n || m_n || i)) mod n */
+    for (i = 0; i < n_sigs; i++) {
+        unsigned char pk_ser[32];
+
+        /* r_i = sig_i[0:32] */
+        if (sig64 != NULL) {
+            secp256k1_sha256_write(&hash, sig64[i], 32);
+        } else {
+            /* aggsig != NULL */
+            secp256k1_sha256_write(&hash, &aggsig[i*32], 32);
+        }
+        if (!secp256k1_xonly_pubkey_serialize(ctx, pk_ser, &pubkey[i])) {
+            return 0;
+        }
+        secp256k1_sha256_write(&hash, pk_ser, sizeof(pk_ser));
+        secp256k1_sha256_write(&hash, msg32[i], 32);
+    }
+    /* TODO: copy midstate instead of computing intermediate "midhash" */
+    secp256k1_sha256_finalize(&hash, midhash);
+    return 1;
+}
+
+int secp256k1_schnorrsig_aggregate(const secp256k1_context* ctx, unsigned char* aggsig, size_t* aggsig_size, unsigned char **sig64, unsigned char **msg32, secp256k1_xonly_pubkey *pubkey, uint32_t n_sigs) {
+    uint32_t i;
+    secp256k1_sha256 hash;
+    unsigned char midhash[32];
+    secp256k1_scalar s;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(aggsig != NULL);
+    ARG_CHECK(aggsig_size != NULL);
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(pubkey != NULL);
+
+    if (*aggsig_size < 32*(1 + n_sigs)) {
```

Comment:
Overflow protection needed here, AFAICT.  You probably want to divide aggsig_size, check it's > 0, then sub 1 and compare with n_sigs.